File: kth_smallest_element_in_a_bst.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} k
# @return {Integer}
def kth_smallest(root, k)
    @count = 0

    kth_smallest_root = return_kth_smallest(root, k)
    return kth_smallest_root.val
end

def return_kth_smallest(root, k)
    if root == nil
        return nil
    end

    left = return_kth_smallest(root.left, k)

    if left != nil
        return left
    end

    # current
    @count += 1
    if @count == k
        return root
    end

    return return_kth_smallest(root.right, k)
end

File: binary_tree/kth_smallest_element_in_a_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        order = []
        # https://www.geeksforgeeks.org/tree-traversals-inorder-preorder-and-postorder/
        # 1. inOrder traversal で配列に格納していく
        # 2. 順番が正しくなかったら、BSTの条件に反するので、return Falseすればいいだけ
        def inOrder(root, order):
            # if not root.val にすると、`0` も False 扱いになるので注意
            if root is None or root.val is None:
                return
            inOrder(root.left, order)
            order.append(root.val)
            inOrder(root.right, order)

        inOrder(root, order)
        # print(order)

        return order[k - 1]
